controllers/vspherecluster_reconciler.go

```diff
@@ -359,7 +359,12 @@ func (r *clusterReconciler) reconcileDeploymentZones(ctx context.Context, cluste
 }
 
 func (r *clusterReconciler) reconcileClusterModules(ctx context.Context, clusterCtx *capvcontext.ClusterContext) (reconcile.Result, error) {
-	if feature.Gates.Enabled(feature.NodeAntiAffinity) {
+	// We want to enable cluster module only when NodeAntiAffinity flag is true and failure domains are not used.
+	// When failure domain is used and an empty value is passed for data center or the resource pool in a vm template,
+	// cluster module creation will fail as there are no values in dc or resource pool.
+	// These values in the vm template gets replaced at a later stage by the values in the failure domain.
+	// When failure domain is used, we do not need the cluster module
+	if feature.Gates.Enabled(feature.NodeAntiAffinity) && clusterCtx.VSphereCluster.Spec.FailureDomainSelector == nil {
```

Comment:
You note:

> [...] During the creation of cluster module, it validates the values in the vm template like data-center, resource-pool etc. When a failure domain is used, these values in the vm template can be empty and it get replaced by the values in the failure domain/deployment zone at a later stage. So, when the value of data-center, resource-pool etc are empty in vm template, the cluster module creation fails and cluster module doesn't make sense when failure domains are being used. To mitigate this issue, we have introduced a check in the vSphere cluster reconciler to enable cluster module creation only when the node anti-affinity is enabled and the failure domains are not being used

Is it **always** the case that cluster modules don't make sense or are there cases where they make sense?

Reply:
@chrischdi,
when failure domain is used it is not making sense as the cluster module creation happens much before the vm template override function. The override function will replace the values in vm template with values in the failure domain. So control plane vms can end up in different zones(vSphere cluster - not vSphereCluster CR) with different values for resource pool, vm folder etc whereas cluster module gets created from the template with wrong values for resource pool, folder etc.

When failure domains are not used, cluster module makes sense

Reply:
But it still makes sense if the different zones are in the same vcenter right?

Reply:
@RenilRaj-BR it feels like this change is orthogonal to adding support for Storage Policies per Failure Domain.

Please consider to move this to a separate PR/issue so we can discuss on this topic properly.

Reply:
sure. I will move it out